File: py/1239.maximum-length-of-a-concatenated-string-with-unique-characters.py
#
# @lc app=leetcode id=1239 lang=python3
#
# [1239] Maximum Length of a Concatenated String with Unique Characters
#
from _typeshed import Self
from typing import *
# @lc code=start


class Solution:
    def maxLength(self, arr: List[str]) -> int:
        def dfs(res: List[int], ls: Optional[List[int]], current: int, pos: int):
            res[0] = max(res[0], bin(current).count('1'))
            if pos == len(ls):
                return
            for i in range(pos, len(ls)):
                if current & ls[i] == 0:
                    dfs(res, ls, current | ls[i], i+1)
        ls = []
        for s in arr:
            val = 0
            for c in s:
                if (val & 1 << ord(c) - ord('a')) == 0:
                    val |= 1 << ord(c)-ord('a')
                else:
                    val = 0
                    break
            if val != 0:
                ls.append(val)
        res = [0]
        dfs(res, ls, 0, 0)
        return res[0]


# @lc code=end

File: cpp/1239.maximum-length-of-a-concatenated-string-with-unique-characters.cpp
/*
 * @lc app=leetcode id=1239 lang=cpp
 *
 * [1239] Maximum Length of a Concatenated String with Unique Characters
 */
#include <vector>
#include <string>
#include <bitset>
#include <set>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
    int vec_len;
public:
    int maxLength(vector<string> &arr)
    {
        vector<bitset<26>> v;
        for (auto &s : arr)
        {
            if (!wordToInt(s).none())
            {
                v.push_back(wordToInt(s));
            }
        }
        int res = 0;
        vec_len=v.size();
        dfs(res, v, bitset<26>(), 0);
        return res;
    }
    void dfs(int &res, vector<bitset<26>> &vec, bitset<26> current, int pos)
    {
        res = max(res, static_cast<int>(current.count()));
        if (pos == vec_len)
        {
            return;
        }
        for (int i = pos; i != vec_len; ++i)
        {
            if (!(current & vec[i]).any())
            {
                dfs(res, vec, current | vec[i], i + 1);
            }
        }
    }
    bitset<26> wordToInt(string &s)
    {
        bitset<26> res;
        for (auto c : s)
        {
            if (!res.test(c - 'a'))
            {
                res.set(c - 'a');
            }
            else
            {
                return bitset<26>();
            }
        }
        return res;
    }
};
// @lc code=end
